### ISSUE TITLE
Bug in NAMD patch

### ISSUE THREAD

**Issue by GiovanniBussi** [plumed/plumed2#254]
It looks like commit 5952dbec4f4f043dcd69c9bc6c47193d0adbbb3b introduced a bug in NAMD patch. This is possible since I am afraid the functions used from NAMD are not included in regtest. I will try to add a suitable regtest  then perhaps I will need @carlocamilloni help to find the problem

See also [this](5952dbec4f4f043dcd69c9bc6c47193d0adbbb3b) post on the mailing list

**Comment by HanatoK** [comment]
I have run rt-make-namd with the following changes:
1. Without fix-254 and without reverting fix 241, cc shows:
```
#! FIELDS time d
 0.000000 12.206556
 0.002000 15.652476
 0.004000 19.104973
 0.006000 22.561028
 0.008000 26.019224
```
2. Without fix-254 and revert fix 241, cc shows:
```
#! FIELDS time d
 0.000000 5.196152
 0.002000 5.196152
 0.004000 5.196152
 0.006000 5.196152
 0.008000 5.196152
```
3. With fix-254, cc shows:
```
#! FIELDS time d
 0.000000 5.196152
 0.002000 3.464102
 0.004000 6.928203
 0.006000 10.392305
 0.008000 13.856406
```

In the regtest, case 3 seems correct, but when using fix-254, the COLVAR output in real namd is just like case 2 (distance in COLVAR never changes).
Case 2 seems wrong but in namd it works.

**Comment by HanatoK** [comment]
@GiovanniBussi 
I have tested the fix-254 branch and the regtest `rt-make-namd`. It's very strange that the colvar output depends on the atoms selected. For `ATOMS=2,3`, it seems work, but for `ATOMS=3,4`, it doesn't.
Could you have a look at the tentative fix and rerun the `rt-make-namd` with `d: DISTANCE ATOMS=3,4`?

**Comment by HanatoK** [comment]
@GiovanniBussi
I don't know which version of `MDAtomsTyped<T>::getPositions` is invoked in other MD engines, but if I change the `MDAtomsTyped<T>::getPositions(const std::set<AtomNumber>&index,const vector<unsigned>&i, vector<Vector>&positions)` to the following together with fix-254:

``` 
template <class T>
void MDAtomsTyped<T>::getPositions(const std::set<AtomNumber>&index,const vector<unsigned>&i, vector<Vector>&positions)const {
// cannot be parallelized with omp because access to positions is not ordered
  unsigned k=0;
  for(const auto & p : index) {
//     positions[p.index()][0]=px[stride*i[k]]*scalep;
//     positions[p.index()][1]=py[stride*i[k]]*scalep;
//     positions[p.index()][2]=pz[stride*i[k]]*scalep;
    positions[p.index()][0]=px[stride*k]*scalep;
    positions[p.index()][1]=py[stride*k]*scalep;
    positions[p.index()][2]=pz[stride*k]*scalep;
    k++;
  }
}
```

NAMD is working correctly! (Although I don't know why it works)

**Comment by GiovanniBussi** [comment]
@HanatoK I had a quick look at the code now and I am confused - sorry I don't remember the regtest.

1. The regtest has no reference result
2. Looking at the code in the regtest I think that the distances **should be constant**! Perhaps it's not well designed. Instead, the positions should vary (check tmp/xyz file).
3. Given this, I think that in your list case 1 is correct and case 2 and 3 are both incorrect.

Bottom line, I suspect the bug introduced by the optimization in atom sharing is not solved yet..

Am I correct?

**Comment by GiovanniBussi** [comment]
@HanatoK notice that positions are set using this instruction:
````
  for(unsigned i=0;i<natoms;i++) all_positions[i]=i+step;
````

Thus pairwise distances should be the same at every step. By the way, there is a stupid mistake, it should have been 3*natoms. Anyway, this only matters for atoms with index > 100/3=33 I think

**Comment by GiovanniBussi** [comment]
@HanatoK can you try the following change (on top of branch fix-254):

````
index e7c025902..872460314 100644
--- a/src/core/Atoms.cpp
+++ b/src/core/Atoms.cpp
@@ -468,7 +468,9 @@ void Atoms::createFullList(int*n) {
     fullList.resize(natoms);
     for(unsigned i=0;i<natoms;i++) fullList[i]=i;
   } else {
-    std::set<AtomNumber> unique;
     for(unsigned i=0; i<actions.size(); i++) {
       if(actions[i]->isActive()) {
         if(!actions[i]->getUnique().empty()) {
````

that is, just remove one line so that the `unique` list used here is `Atoms::unique` instead of being a local variable. This is the same list that is later used to move around atoms in plumed. Let me know if this fixes the problem with real NAMD. In case so, I will update the regtest. Thanks!

**Comment by GiovanniBussi** [comment]
You should also clear the unique container (with `unique.clear()`).

I actually pushed a new commit on the fix-254 branch right now.

**Comment by GiovanniBussi** [comment]
Sorry, I think there are still problems...

I have no idea how to solve them right now.

**Comment by HanatoK** [comment]
1. Without `unique.clear()` the NAMD and the xyz of regtest looks working, but all atoms are requested at every step (I is always 100). This may hurt the performance.
2. With `unique.clear()`  the NAMD doesn't update the COLVAR, and also the xyz of regtest isn't updated. Only at first step 100 atoms are requested (as the I in std::err), 2 atoms (defined in the distance colvar) are requested at following steps.

**Comment by HanatoK** [comment]
Currently my solution is modifying `void MDAtomsTyped<T>::getPositions(const std::set<AtomNumber>&index,const vector<unsigned>&i, vector<Vector>&positions)` in MDAtoms.cpp. This function is introduced in #fix-241. The original code is commented out.
```
template <class T>
void MDAtomsTyped<T>::getPositions(const std::set<AtomNumber>&index,const vector<unsigned>&i, vector<Vector>&positions)const {
// cannot be parallelized with omp because access to positions is not ordered
  unsigned k=0;
  for(const auto & p : index) {
//     positions[p.index()][0]=px[stride*i[k]]*scalep;
//     positions[p.index()][1]=py[stride*i[k]]*scalep;
//     positions[p.index()][2]=pz[stride*i[k]]*scalep;
    positions[p.index()][0]=px[stride*k]*scalep;
    positions[p.index()][1]=py[stride*k]*scalep;
    positions[p.index()][2]=pz[stride*k]*scalep;
    k++;
  }
}
```
This can make both real NAMD and the regtest working.

**Comment by GiovanniBussi** [comment]
Yes that's exactly the path taken in NAMD patch.

Line 201: This is for codes asking all the atoms.
Lines 206/208: This is for codes asking only the atoms in the
std:set<AtomNumber> called unique.

Line 206 is used when using domain decomposition or when, even with a
single domains, atoms are reordered (the latter is LAMMPS)
Line 208 is used in NAMD.

uniq_index[i] is expected to be absolute index of the i-th atom in the
array passed by the MD code.

I think that line 208 is incorrect.

Notice that dd.g2l should be correctly defined also with NAMD. Indeed from
the patch file I can see that NAMD does this:
- ask the list to PLUMED
- collect the atoms
- pass the list again to PLUMED (cmd("setAtomsGatindex",&index[0]);).


Maybe if you just change

if(dd && shuffledAtoms>0) {
      for(const auto & p : unique) uniq_index.push_back(dd.g2l[p.index()]);
    } else {
      for(const auto & p : unique) uniq_index.push_back(p.index());
    }

to

for(const auto & p : unique) uniq_index.push_back(dd.g2l[p.index()]);

everything will work?

Giovanni

**Comment by HanatoK** [comment]
The domain decomposition doesn't work in both NAMD and the regtest case. Using `for(const auto & p : unique) uniq_index.push_back(dd.g2l[p.index()]);` directly will cause a segfault.

**Comment by HanatoK** [comment]
@GiovanniBussi 
I have tried introducing a new vector called g2l_nondd which is similar to the dd.g2l in the Atoms class. In NAMD and the regtest case it works and it passes most of the tests in travis-CI. Is this the right way to fix this bug?

**Comment by HanatoK** [comment]
@GiovanniBussi 
I have made a patch and tested it on the travis-CI (https://travis-ci.org/HanatoK/plumed2/builds/369918959?utm_source=github_status&utm_medium=notification). This patch moves the g2l vector out of the dd class.
Still, I am not familiar with the PLUMED internal code and data structures so I couldn't guarantee it always works.
The patch is attached as following:
[254.txt](https://github.com/plumed/plumed2/files/1936485/254.txt)

**Comment by GiovanniBussi** [comment]
Fixed (see #359).
